File: solutions/scala/152-combination-sum-iv.scala
/**
 * Problem 152: Combination Sum IV (LeetCode 377)
 * Difficulty: Med
 * Language: Scala
 */
def combinationSum4(nums: Array[Int], target: Int): Int =
  (1 to target).foldLeft(Array.fill(target + 1)(0).updated(0, 1)) { (dp, i) =>
    dp.updated(i, nums.filter(_ <= i).map(n => dp(i - n)).sum)
  }(target)

File: solutions/typescript/152-combination-sum-iv.ts
/**
 * Problem 152: Combination Sum IV (LeetCode 377)
 * Difficulty: Med
 * Language: TypeScript
 */
function combinationSum4(nums: number[], target: number): number {
  const dp = new Array(target + 1).fill(0);
  dp[0] = 1;
  return Array.from({ length: target }, (_, i) => i + 1).reduce((dp, i) => {
    dp[i] = nums.reduce((sum, n) => sum + (i >= n ? dp[i - n] : 0), 0);
    return dp;
  }, dp)[target];
}
